### ISSUE TITLE
Problem with getting an admin token (Ubuntu on WSL)

### ISSUE THREAD

**Issue by LukasPaschen** [clp-research/slurk#99]
After the execution of get_admin_token.sh, the environment variable $ADMIN_TOKEN is still empty. Docker seems to be installed correctly and start_slurk_server.sh seems to work as expected ($SLURK_SERVER_ID is not empty after execution). 

`docker logs $SLURK_SERVER_ID`
returns
```
[2020-09-21 13:51:18 +0000] [1] [INFO] Starting gunicorn 19.9.0
[2020-09-21 13:51:18 +0000] [1] [INFO] Listening at: http://0.0.0.0:5000 (1)  
[2020-09-21 13:51:18 +0000] [1] [INFO] Using worker: geventwebsocket.gunicorn.workers.GeventWebSocketWorker  
[2020-09-21 13:51:18 +0000] [9] [INFO] Booting worker with pid: 9  
[2020-09-21 13:51:18 +0000] [9] [ERROR] Exception in worker process  
Traceback (most recent call last):  
  File "/usr/local/lib/python3.7/site-packages/gunicorn/arbiter.py", line 583, in spawn_worker
    worker.init_process()  
  File "/usr/local/lib/python3.7/site-packages/gunicorn/workers/ggevent.py", line 196, in init_process
    self.patch()  
  File "/usr/local/lib/python3.7/site-packages/gunicorn/workers/ggevent.py", line 75, in patch
    fileno=s.sock.fileno()))  
  File "/usr/local/lib/python3.7/site-packages/gevent/_socket3.py", line 114, in __init__
    self._sock = self._gevent_sock_class(family, type, proto, fileno)  
OSError: [Errno 92] Protocol not available: 'protocol'  
[2020-09-21 13:51:18 +0000] [9] [INFO] Worker exiting (pid: 9)  
[2020-09-21 13:51:18 +0000] [1] [INFO] Shutting down: Master  
[2020-09-21 13:51:18 +0000] [1] [INFO] Reason: Worker failed to boot.
```
This might have something to do with using Ubuntu on Windows Subsystem for Linux, but I am not sure. Many thanks in advance.

**Comment by phisad** [comment]
Not sure if slurk really started up correctly as the `docker logs` should show the admin token. Could you please provide a step-by-step guide of how you started slurk (so I can reproduce this). 

Which docker version? Is "Use the WSL 2 based engine" checked?

**Comment by LukasPaschen** [comment]
First of all, thanks for your quick response!
My docker version is ```18.06.1-ce```. I think "Use the WSL 2 based engine" is an option in Docker Desktop, right? I do not use Docker Desktop, because I have the wrong Windows version and for some docker versions it is possible to install them without Docker Desktop. (I installed docker following https://stackoverflow.com/a/56051167.) I think docker is installed correctly, ```docker run hello-world``` works as expected.
I started slurk with these commands:
```source scripts/start_slurk_server.sh```
This returns ```slurky``` as expected. But when I run ```source scripts/get_admin_token.sh```, it does not create a token.

**Comment by phisad** [comment]
Well, the `get_admin_token.sh` is **not** creating an admin token, but fetching it from the `docker logs` command. The usual output looks like:

```
PS F:\Development\git\slurk> docker logs slurky
[2020-09-22 16:18:24 +0000] [1] [INFO] Starting gunicorn 19.9.0
[2020-09-22 16:18:24 +0000] [1] [INFO] Listening at: http://0.0.0.0:5000 (1)
[2020-09-22 16:18:24 +0000] [1] [INFO] Using worker: geventwebsocket.gunicorn.workers.GeventWebSocketWorker
[2020-09-22 16:18:24 +0000] [8] [INFO] Booting worker with pid: 8
admin token:
ef2a9505-67d6-4b45-9371-a89ff448ca60
```

Nevertheless, as it seems, you were not able to start _slurk_ in the first place using the docker installation of the WSL.

As a workaround, you could try to install Docker Desktop for Windows 10. Are you missing the Pro Version for Hyper-V? Even better would be to install a Linux system or use a remote machine on which Linux is installed.

Furthermore, I will try to install docker using the WSL, but I think this will likely only lead to your error above.

**Comment by phisad** [comment]
@LukasPaschen I guess there is good and bad news:
- Good: I could not reproduce your issee
- Bad: I used slightly different versions

**TL;DR**
Please try to install the following versions (see instructions below):
- OS Name:                   Microsoft Windows 10 Pro
- OS Version:                10.0.19041 N/A Build 19041
- WSL2
- Ubuntu for WSL 20.04.1
- docker ce 19.03.13

I followed the official documentations (I also used the new Windows Terminal (which is quite awesome)):

**Setting up WSL 2 on Windows 10**
https://docs.microsoft.com/de-de/windows/wsl/install-win10

1. Open a Powershell as Administrator
1. $> dism.exe /online /enable-feature /featurename:Microsoft-Windows-Subsystem-Linux /all /norestart
1. $> dism.exe /online /enable-feature /featurename:VirtualMachinePlatform /all /norestart
1. $> wsl --set-default-version 2
1. Restart the system

**Install Ubuntu-20.04.1 using App Store** 

1. Click "Launch" after install 
1. Setting Unix username and password
1. Open a Powershell
1. Check the used WSL version for each distribution using PS: 
$> wsl --list --verbose
```  
  NAME                   STATE           VERSION
  Ubuntu-20.04           Running         2
```

**Install docker into the Ubuntu WSL**
https://docs.docker.com/engine/install/ubuntu/

1. Open a shell for the Ubuntu sub-system
1. Uninstall old versions: 
$> sudo apt-get remove docker docker-engine docker.io containerd runc
1. Install using the repository: 
$> sudo apt-get update
$> sudo apt-get install apt-transport-https ca-certificates curl gnupg-agent software-properties-common
    
1. Add Docker’s official GPG key:
$> curl -fsSL https://download.docker.com/linux/ubuntu/gpg | sudo apt-key add -

1. Use the following command to set up the stable repository.
$> sudo add-apt-repository "deb [arch=amd64] https://download.docker.com/linux/ubuntu $(lsb_release -cs) stable"
   
1. Install the docker engine
$> sudo apt-get update
$> sudo apt-get install docker-ce docker-ce-cli containerd.io   

**Potential after-installation steps:**
https://docs.docker.com/engine/install/linux-postinstall/

1. Run the service in the Ubuntu WSL:
$> sudo service docker status
$> sudo service docker start

1. Add the non-root user to the docker group and re-login:
$> sudo groupadd docker
$> sudo usermod -aG docker ${USER}
$> su -s ${USER}

**Checkout and run slurk**
Notice: We need to checkout the project using the WSL otherwise the line-endings are not correct!

1. $> ssh-keygen
1. Copy the generated public key to your github settings
1. $> git clone git@github.com:clp-research/slurk.git
1. Go into the slurk top directory
$> source ./start_slurk_server.sh
$> echo $SLURK_SERVER_ID
ce96a2009c0be81f6a82161f611db2f5fd74ed95e8d9b8ea434aeb1bcbb2342b
$> source ./get_admin_token.sh
$> echo $ADMIN_TOKEN
d858f508-da42-4ffb-847c-c7f378c35e02

**Side-notes:**

The installation of the docker-ce showed some "failures", but these seem not to have an impact on the docker engine.

```
Setting up docker-ce-cli (5:19.03.13~3-0~ubuntu-focal) ...
Setting up pigz (2.4-1) ...
Setting up cgroupfs-mount (1.4) ...
invoke-rc.d: could not determine current runlevel
Setting up grub2-common (2.04-1ubuntu26.4) ...
Setting up grub-pc-bin (2.04-1ubuntu26.4) ...
Setting up docker-ce (5:19.03.13~3-0~ubuntu-focal) ...
Created symlink /etc/systemd/system/multi-user.target.wants/docker.service → /lib/systemd/system/docker.service.
Created symlink /etc/systemd/system/sockets.target.wants/docker.socket → /lib/systemd/system/docker.socket.
invoke-rc.d: could not determine current runlevel
Setting up grub-pc (2.04-1ubuntu26.4) ...

Creating config file /etc/default/grub with new version
Setting up grub-gfxpayload-lists (0.7) ...
Processing triggers for install-info (6.7.0.dfsg.2-5) ...
Processing triggers for libc-bin (2.31-0ubuntu9) ...
Processing triggers for systemd (245.4-4ubuntu3.2) ...
Processing triggers for man-db (2.9.1-1) ...
Processing triggers for linux-image-unsigned-5.6.0-1028-oem (5.6.0-1028.28) ...
/etc/kernel/postinst.d/initramfs-tools:
update-initramfs: Generating /boot/initrd.img-5.6.0-1028-oem
W: mkconf: MD subsystem is not loaded, thus I cannot scan for arrays.
W: mdadm: failed to auto-generate temporary mdadm.conf file.
```
For details see:
> https://github.com/microsoft/WSL/issues/4903 (mdadm 4.1-2 causes unrecoverable grub config loop in dpkg, apt)
> Spiritually similar #4763 (over there it was LVM). WSL2 does not load the Linux kernel with GRUB (and WSL1 doesn't load a Linux kernel at all). Linux RAID (mdadm) isn't applicable for lack of Multiple Disks. Somehow those were installed on your system. Purging those packages should help.

**Comment by janagoetze** [comment]
@LukasPaschen you can find the additional documentation in the docs now: https://clp-research.github.io/slurk/slurk_prerequisites.html
